### PR TITLE
Use MSBuild /restore

### PR DESCRIPTION
**Pull request by jnm2**
Simplifies by using MSBuild's `/restore` parameter (Cake 0.27's `WithRestore()`) which runs the Restore target in its own MSBuild context before running the specified target.

(Been around for nearly a year: `https://github.com/Microsoft/msbuild/pull/2414`)

### CODE REVIEW

File: build.cake [hunk 1]
```diff
@@ -158,41 +144,29 @@ Task("UpdateAssemblyInfo")
 // BUILD ENGINE
 //////////////////////////////////////////////////////////////////////
 
+MSBuildSettings CreateMSBuildSettings(string target) => new MSBuildSettings()
+    .SetConfiguration(configuration)
+    .SetVerbosity(Verbosity.Minimal)
+    .WithProperty("PackageVersion", productVersion)
+    .WithTarget(target);
```
Comment (jnm2) [inline build.cake]: ℹ This helper method makes it easy to drop in workarounds like `.WithProperty("AddSyntheticProjectReferencesForSolutionDependencies", "false")` in a single place, which I need right now to work around `https://github.com/Microsoft/msbuild/issues/3626`.

File: build.cake [hunk 2]
```diff
@@ -206,22 +180,16 @@ Task("BuildNetStandardEngine")
...
-            .SetVerbosity(Verbosity.Minimal)
-            .SetNodeReuse(false)
-        );
+        MSBuild(
+            "./src/NUnitEngine/mock-cpp-clr/mock-cpp-clr-x86.vcxproj",
+             CreateMSBuildSettings("Build").WithProperty("Platform", "x86"));
+
+        MSBuild(
+            "./src/NUnitEngine/mock-cpp-clr/mock-cpp-clr-x64.vcxproj",
+            CreateMSBuildSettings("Build").WithProperty("Platform", "x64"));
```
[the remark this answers is not in the document]
Comment (jnm2) [reply build.cake]: If I understand correctly, node reuse only happens if you pass `/m` to do a multi-process build, which in Cake looks like `.SetMaxCpuCount(null)`.
`https://github.com/Microsoft/msbuild/issues/2916#issuecomment-360897413`

File: build.cake [hunk 3]
```diff
@@ -158,41 +144,29 @@ Task("UpdateAssemblyInfo")
...
+        if (IsDotNetCoreInstalled)
         {
-            var settings = new DotNetCoreBuildSettings
-            {
-                Configuration = configuration,
-                EnvironmentVariables = new Dictionary<string, string>()
-            };
-            settings.EnvironmentVariables.Add("PackageVersion", productVersion);
-            DotNetCoreBuild(DOTNETCORE_SOLUTION_FILE, settings);
+            MSBuild(DOTNETCORE_SOLUTION_FILE, CreateMSBuildSettings("Build").WithRestore());
```
[the remark this answers is not in the document]
Comment (jnm2) [reply build.cake]: As far as I know the only thing the `PackageVersion` env var affects is that it env vars become MSBuild global properties. So this became `MSBuildSettings.WithProperty` to have the same global proeprty effect on MSBuild via the command line.

**Comment by jnm2** [comment]
@mikkelbu 

> I've some comments/questions in the code, but I'm also wondering if this sets any restriction on the version of VS2017 that one is using. E.g. if one has a (very) old installation with an old msbuild (I guess that msbuild is shipped with VS, but I'm unsure) then this would break in his/her's setup?

Very good question. The solution already requires Visual Studio 2107 (15.x) to build. According to `https://github.com/dotnet/cli/pull/7693`, `/restore` was first available in MSBuild 15.5. I think MSBuild always aligns with Visual Studio version numbers, so that would indicate that the first release version of Visual Studio that has `/restore` is Visual Studio 2017 Update 5, shipped 2017-12-04. (On my birthday 😁)

I'm personally in favor of supporting one or more specific major versions of IDEs but I'm not really in favor of supporting very old minor or patch versions within those major version trains. It's not likely that anyone building this repo from source has a VS2017 installation that they haven't updated for over nine months, and if so, requiring a minor version update to be within the last nine months doesn't seem over-the-top to me.

Great topic and that's my personal opinion, but I'm happy to abide by the wishes of the @nunit/engine-team.

**Comment by jnm2** [comment]
Another thought: while working on removing Mono.Cecil, I keep trying to use C# 7.2 syntax like target-typed `default` and `readonly struct`. C# 7.2 requires the same VS2017 Update 5 release.

**Comment by jnm2** [comment]
@rprouse ...wow okay! 😃 (updated)

**Comment by rprouse** [comment]
@jnm2 I've approved this, but the AppVeyor build is failing for the same reasons as I am seeing trying to build master in prep for the release. I am also seeing that failure on my "merge .NET Standard back into the main engine" branch for awhile now and have been pulling my hair out trying to solve it. I thought it was something I changed, but now I see it is something that changed on my computer.

Obviously you are building successfully from the command line, so I am trying to figure out what is different. From what I can see;

- Building from Visual Studio works for me
- AppVeyor and I are both using .NET Core SDK 2.1.401 although I had the same problem with earlier 2.1 versions. I assume you have 2.1 installed?
- My MSBuild Engine reports as 15.8.166, AppVeyor is 15.8.168. It has been happening to me since at least 15.7 though.
- Last successful build on AppVeyor was 11 days ago. AppVeyor updated to 15.8.1 and then 15.8.2 in that time, but this was happening to me on older versions. https://www.appveyor.com/updates/

The error is;

```
========================================
BuildNetFramework
========================================
Executing task: BuildNetFramework
Microsoft (R) Build Engine version 15.8.166+gd4e8d81a88 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  nunit.engine.api -> C:\src\nunit\nunit-console\bin\Release\net20\nunit.engine.api.dll
  nunit.engine -> C:\src\nunit\nunit-console\bin\Release\net20\nunit.engine.dll
  nunit-agent -> C:\src\nunit\nunit-console\bin\Release\net20\nunit-agent.exe
  nunit3-console -> C:\src\nunit\nunit-console\bin\Release\net20\nunit3-console.exe
  nunit-agent-x86 -> C:\src\nunit\nunit-console\bin\Release\net20\nunit-agent-x86.exe
  mock-assembly -> C:\src\nunit\nunit-console\bin\Release\net35\mock-assembly.dll
  mock-assembly -> C:\src\nunit\nunit-console\bin\Release\netstandard1.6\mock-assembly.dll
  notest-assembly -> C:\src\nunit\nunit-console\bin\Release\net35\notest-assembly.dll
C:\Program Files\dotnet\sdk\2.1.401\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(198,5): error NETSDK1005: Assets file 'C:\src\nunit\nunit-console\src\NUnitEngine\nunit-agent\obj \project.assets.json' doesn't have a target for '.NETFramework,Version=v3.5'. Ensure that restore has run and that you have included 'net35' in the TargetFrameworks for your project. [C:\src\nunit\nunit-console \src\NUnitEngine\nunit-agent\nunit-agent.csproj]
C:\Program Files\dotnet\sdk\2.1.401\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(198,5): error NETSDK1005: Assets file 'C:\src\nunit\nunit-console\src\NUnitEngine\nunit-agent\obj \project.assets.json' doesn't have a target for '.NETFramework,Version=v3.5'. Ensure that restore has run and that you have included 'net35' in the TargetFrameworks for your project. [C:\src\nunit\nunit-console \src\NUnitEngine\nunit-agent\nunit-agent-x86.csproj]
C:\Program Files\dotnet\sdk\2.1.401\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(198,5): error NETSDK1005: Assets file 'C:\src\nunit\nunit-console\src\NUnitConsole\nunit3-console \obj\project.assets.json' doesn't have a target for '.NETFramework,Version=v3.5'. Ensure that restore has run and that you have included 'net35' in the TargetFrameworks for your project. [C:\src\nunit\nunit-con sole\src\NUnitConsole\nunit3-console\nunit3-console.csproj]
```

Any ideas would be appreciated. I'm going to try on my work computer in the morning and compare configurations.

**Comment by rprouse** [comment]
I don't think it is the version of the .NET Core SDK that is installed. I dropped a `global.json` in the root and pegged to 1.0.4 and 2.0.3 and still get the same error (after deleting the `obj` directories to be sure). I am also getting this error with a clean clone of the repo.

**Comment by rprouse** [comment]
Nice, that was quick. Either you have great Google skills or a great memory!

**Comment by jnm2** [comment]
@rprouse It took me a second to get my head back in the game, but I hit it here: https://github.com/nunit/nunit-console/issues/389#issuecomment-415916343

**Pull request merged**